### ISSUE TITLE
Wzero-as-null-pointer-constant warns when using `operator<=>`

### ISSUE THREAD

**Issue by davidstone** [llvm/llvm-project#43670]
|  |  |
| --- | --- |
| Bugzilla Link | [44325](https://llvm.org/bz44325) |
| Version | unspecified |
| OS | Linux |
| CC | @adrianimboden,@dwblaikie,@0x8000-0000,@dvhwgumby,@mclow,@mwinterb,@mizvekov,@zygoloid,@Weverything |
| Fixed by commit(s) | 1f06f41993b6363e6b2c4f22a13488a3e687f31b, 678c259d277135ef32861887a8ac8618deba5f24 |

## Extended Description 
```
#include <compare>

auto b = 1 <=> 2 < 0;
```

Outputs:

```
<source>:3:20: warning: zero as null pointer constant [-Wzero-as-null-pointer-constant]

auto b = 1 <=> 2 < 0;

                   ^

                   nullptr

1 warning generated.

Compiler returned: 0
```

See it live: https://godbolt.org/z/9a8cnD

Technically, the warning is accurate. User code has a literal `0` that is being passed to a function accepting a `nullptr`. However, this is an implementation detail of the standard library and the intended interface for use (users should not use `nullptr` there), so we should not warn users for doing so (and definitely should not direct them to change their code).

**Comment by 0x8000-0000** [comment]
The problem might not be in libc++, because


```c++
#include <chrono>

using DeviceTimePoint = std::chrono::microseconds;

bool busyLoop(std::chrono::microseconds timeOut);

DeviceTimePoint getTime();

bool isBusy();

bool busyLoop(std::chrono::microseconds timeOut)
{
    bool status = true;
    const auto stopTime = getTime() + timeOut;

    while (isBusy())
    {
        if (getTime() >= stopTime)
        {
            status = false;
            break;
        }
    }
    return status;
}
```

Produces this:

```console
$ /opt/clang11/bin/clang++ --gcc-toolchain=/opt/gcc10 -std=c++20 -c -Werror -Wzero-as-null-pointer-constant compare_durations.cpp
compare_durations.cpp:18:23: error: zero as null pointer constant [-Werror,-Wzero-as-null-pointer-constant]
        if (getTime() >= stopTime)
                      ^~
                      nullptr
/opt/gcc10/lib/gcc/x86_64-pc-linux-gnu/10.1.0/../../../../include/c++/10.1.0/chrono:677:7: note: while rewriting comparison as call to 'operator<=>' declared here
      operator<=>(const duration<_Rep1, _Period1>& __lhs,
      ^
1 error generated.
```

**Comment by DeadlyRedCube** [comment]
In Clang 14.0.1 I'm getting this without an explicit comparison against 0, using an auto-generated operator<=>:

```
#include <chrono>

class Wrapper
{
public:
    std::strong_ordering operator<=>(const Wrapper &other) const noexcept = default;

private:
    std::chrono::system_clock::time_point t;
};



bool Foo()
{
    Wrapper a, b;
    return a < b;
}
```

This code doesn't have any explicit 0 checks in it, it's all in the autogenerated code (here's a compiler explorer link, if that helps:
  [https://godbolt.org/z/jWxhdMz8a](https://godbolt.org/z/jWxhdMz8a). 

I ended up working around it by writing a custom <=> (that just returns (t <=> other.t)) and then a default operator==, which works but definitely isn't ideal.

**Comment by frederick-vs-ja** [comment]
It's unfortunate that we need to either accept `1 - 1` or treat `0` as a null pointer constant. Perhaps we should ask the Clang frontend to add an internal type which can only be initialized from literal `0`.

**Comment by mordante** [comment]
I tested this with clang-tidy-18 recently and it didn't issue a diagnostic about this. Can you test with clang-18?

**Comment by frederick-vs-ja** [comment]
The warning is still triggered with options `-std=c++20 -stdlib=libc++ -Wzero-as-null-pointer-constant` ([Godbolt link](https://godbolt.org/z/hsheejdjx)).

**Comment by cor3ntin** [comment]
@mordante @ldionne As unpalatable as i think it is, adding an attribute in the front-end might be the most robust solution, it it helps and can be done, ABI wise along the line of `void f(int [[clang::literal_zero]]  i);`.
Let me know what you think, i might have time for clang 19

**Comment by ldionne** [comment]
@cor3ntin I'm trying something out in #79465 . Please let me know what you think about it. We could indeed use `[[clang::literal_zero]]` to make libc++ a bit stricter if we go down the route of my PR.

**Comment by davidstone** [comment]
This still warns with libstdc++: https://godbolt.org/z/hjhWxYeG8

**Comment by cor3ntin** [comment]
@davidstone Well, libstdc++ does not use `enable_if`, so on the clang side the only solution would be to special case the diagnostic emitted from that one function and we aren't too keen on doing that
